### DISCUSSION TITLE
VBR vs VBR Capped

### DISCUSSION THREAD

**Discussion by clee604** [gtxaspec/wz_mini_hacks#281] [Q&A]
Hi, I'm trying to find a little more information on the difference between the different video enc modes:
`ENC_PARAMETER accepts numbers only. 0=FIXQP, 1=CBR, 2=VBR, 4=CAPPED VBR, 8=CAPPED QUALITY. Currently only 2, 4, and 8 are working`

Since there is a target and max bitrate options, what are the differences between VBR, CAPPED VBR and CAPPED QUALITY? Does one prioritize picture quality over framerate? Or does does CAPPED VBR respect the hard limit but VBR doesn't? 

Any info would be greatly appreciated!

**Comment by claudobahn** [comment]
From https://blog.twitch.tv/en/2017/04/04/segment-based-rate-control-of-video-encoder-for-live-abr-streaming-1da5e60a2ec1

![image](https://user-images.githubusercontent.com/2065015/186953339-aedf1e48-b1ac-45e8-a350-5cc3d5f87ee2.png)

The way I understand it:
* mode 2/VBR will make a best effort to target the target bitrate, but will exceed when necessary
* mode 4/Capped VBR will start compromising the video quality to stay under a hard threshold. Not sure if that ends up being the max configured or some other threshold relative to target
* mode 8/Capped Quality AFAIK tries to target a different metric (the QP) which could result in less wasted bitrate and since it's the capped QP version, it has a hard ceiling again. Not really sure how much the max bitrate figures into this mode, or if it's just the max QP that matters

For context, the high res mode by default is configured like so (on my camera, at least) as described by `impdbg --enc_info`:
```
	CHANNEL 0    1920x 1080     START H264 tf:352150     df:2          encdur:175894986, encodingFrameCnt:352148,endencodeFrameCnt=352148,endrelaseFrameCnt=352148,Fps:2.01,Bitrate:872.36
-------------------------------------------------
ch->index = 0
ch->releaseFrmNum = 0
ch->releaseFrmDen = 0
chnAttr->encAttr->eProfile = 77(0x4d) offset:size = 0:4
chnAttr->encAttr->uLevel = 41(0x29) offset:size = 4:1
chnAttr->encAttr->uTier = 0(0x0) offset:size = 5:1
chnAttr->encAttr->uWidth = 1920(0x780) offset:size = 6:2
chnAttr->encAttr->uHeight = 1080(0x438) offset:size = 8:2
chnAttr->encAttr->ePicFormat = 392(0x188) offset:size = 12:4
chnAttr->encAttr->eEncOptions = 294952(0x48028) offset:size = 16:4
chnAttr->encAttr->eEncTools = 156(0x9c) offset:size = 20:4
chnAttr->rcAttr->rcMode = 8(0x8) offset:size = 44:4
chnAttr->rcAttr->CappedQuality->uTargetBitRate = 720(0x2d0) offset:size = 48:4
chnAttr->rcAttr->CappedQuality->uMaxBitRate = 960(0x3c0) offset:size = 52:4
chnAttr->rcAttr->CappedQuality->iInitialQP = -1(0xffffffff) offset:size = 56:2
chnAttr->rcAttr->CappedQuality->iMinQP = 23(0x17) offset:size = 58:2
chnAttr->rcAttr->CappedQuality->iMaxQP = 51(0x33) offset:size = 60:2
chnAttr->rcAttr->CappedQuality->iIPDelta = 3(0x3) offset:size = 62:2
chnAttr->rcAttr->CappedQuality->iPBDelta = 3(0x3) offset:size = 64:2
chnAttr->rcAttr->CappedQuality->eRcOptions = 1(0x1) offset:size = 68:4
chnAttr->rcAttr->CappedQuality->uMaxPictureSize = 1920(0x780) offset:size = 72:4
chnAttr->rcAttr->CappedQuality->uMaxPSNR = 48(0x30) offset:size = 76:2
chnAttr->rcAttr->outFrmRate->frmRateNum = 2(0x2) offset:size = 80:4
chnAttr->rcAttr->outFrmRate->frmRateDen = 1(0x1) offset:size = 84:4
chnAttr->gopAttr->uGopCtrlMode = 2(0x2) offset:size = 88:4
chnAttr->gopAttr->uGopLength = 4(0x4) offset:size = 92:2
chnAttr->gopAttr->uNotifyUserLTInter = 0(0x0) offset:size = 94:1
chnAttr->gopAttr->uMaxSameSenceCnt = 1(0x1) offset:size = 96:4
chnAttr->gopAttr->bEnableLT = 0(0x0) offset:size = 100:1
chnAttr->gopAttr->uFreqLT = 0(0x0) offset:size = 104:4
chnAttr->gopAttr->bLTRC = 0(0x0) offset:size = 108:1
```

It's been a while since I've played w/ it, but in my experience adjusting these two:
```
chnAttr->rcAttr->CappedQuality->iMinQP = 23(0x17) offset:size = 58:2
chnAttr->rcAttr->CappedQuality->iMaxQP = 51(0x33) offset:size = 60:2
```
had a definite impact on the quality/bitrate.

Looking at the SDK sample code:
* VBR: https://github.com/cgrrty/Ingenic-SDK-T31-1.1.1-20200508/blob/3d1ba957605985fd436282fb16de8fcba536a4e9/samples/libimp-samples/sample-common.c#L670
* Capped VBR: https://github.com/cgrrty/Ingenic-SDK-T31-1.1.1-20200508/blob/3d1ba957605985fd436282fb16de8fcba536a4e9/samples/libimp-samples/sample-common.c#L681
* Capped Quality: https://github.com/cgrrty/Ingenic-SDK-T31-1.1.1-20200508/blob/3d1ba957605985fd436282fb16de8fcba536a4e9/samples/libimp-samples/sample-common.c#L693

You can see that Capped VBR and Capped Quality are configured the same way, but the ceiling is different. For Capped VBR the hard ceiling is the bitrate, whereas for Capped Quality the hard ceiling is the quality.

I'm not an expert and don't have the answers, but hopefully this is a step in the right direction.

**Reply by clee604** [reply]
Thanks for the insight! I have been testing option '2' VBR and '4" Capped VBR on my cameras; the main issue I'm trying to deal with is I have some OLD low power devices and they drop/skip some frames at times. 

It generally happens when there is more motion so I assume it's due to increase in bitrate which causes old devices to be unable to cope with the rtsp stream. I rather not split into lower quality streams as it only impacts a few devices, so I've been testing Capped VBR to see if it helps with things. So far I think it does, maybe the normal VBR lets the bitrate exceed the max bitrate target before bringing it back in line, but I'm still testing at this point.
